dali/kernels/common/transpose.h

```diff
@@ -0,0 +1,214 @@
+  int current_start = 0;
+  int current_end = 0;
+  for (int i = 0; i < perm.size() - 1; i++) {
+    if (perm[i] == perm[i + 1] - 1) {
+      current_end = i + 1;
+    } else {
+      result.push_back({current_start, current_end});
+      current_start = i + 1;
+      current_end = i + 1;
+    }
+  }
+  // We wouldn't add the last one in any case
+  result.emplace_back(current_start, current_end);
+  return result;
+}
+
+/**
+ * @brief Based on the blocks returned from PermutationBlocks, collapse the blocks
+ *        in the perm and re-enumerate them.
+ *
+ * @param perm Permutation to collapse
+ * @param perm_blocks Description of consecutive blocks in perm
+ * @return std::vector<int>
+ */
+std::vector<int> CollapsePermutation(span<const int> perm,
+                                     const std::vector<std::pair<int, int>> &perm_blocks) {
+  std::vector<int> result;
+  std::vector<int> removed;
+  result.reserve(perm_blocks.size());
+  removed.reserve(perm.size() - perm_blocks.size());
+  for (auto &block : perm_blocks) {
+    result.push_back(perm[block.first]);
+    for (int i = block.first + 1; i <= block.second; i++) {
+      removed.push_back(perm[i]);
+    }
+  }
+  for (auto &elem : result) {
+    int original_elem = elem;
+    for (auto removed_elem : removed) {
+      if (original_elem > removed_elem) {
+        elem--;
+      }
+    }
+  }
+  return result;
+}
+
+/**
+ * @brief Based on the blocks returned from PermutationBlocks, collapse the blocks
+ *        in the shape that are permuted together.
+ *
+ * @param shape Shape to be collapsed
+ * @param perm Original, non-collapsed permutation
+ * @param perm_blocks Description of blocks in permutation
+ * @return TensorShape<>
+ */
+TensorShape<> CollapseShape(const TensorShape<> &shape, span<const int> perm,
+                            const std::vector<std::pair<int, int>> &perm_blocks) {
+  TensorShape<> result;
+  result.resize(perm_blocks.size());
+  auto collapsed_perm = CollapsePermutation(perm, perm_blocks);
+  for (size_t i = 0; i < perm_blocks.size(); i++) {
+    result[i] = 1;
+  }
+  for (size_t i = 0; i < perm_blocks.size(); i++) {
+    int source_pos = collapsed_perm[i];
+    for (int j = perm_blocks[i].first; j <= perm_blocks[i].second; j++) {
+      result[source_pos] *= shape[perm[j]];
+    }
+  }
+  assert(volume(result) == volume(shape));
+  return result;
+}
+
+}  // namespace detail
+
+/**
+ * @brief Transpose `src` Tensor to `dst` wrt to permutation `perm`
+ */
+template <typename T>
+void Transpose(const TensorView<StorageCPU, T> &dst, const TensorView<StorageCPU, const T> &src,
+               span<const int> perm) {
+  int N = src.shape.sample_dim();
+  assert(dst.shape.sample_dim() == N);
+  assert(volume(src.shape) == volume(dst.shape));
+  auto dst_strides = GetStrides(dst.shape);
+  auto src_strides = GetStrides(src.shape);
+  VALUE_SWITCH(N, static_dims, (1, 2, 3), (
+    detail::transpose_impl_static<T, static_dims, static_dims>(
```

Comment:
If you put dims arguments first, you wouldn't have to specify the type - it could be inferred from arguments.

Reply:
Done